src/components/MainSidebar/ProgramDimensionsPanel/ProgramDimensionsPanel.js

```diff
@@ -0,0 +1,190 @@
+import { useDataQuery } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import { NoticeBox, CenteredContent, CircularLoader } from '@dhis2/ui'
+import cx from 'classnames'
+import PropTypes from 'prop-types'
+import React, { useEffect, useState } from 'react'
+import { useSelector, useDispatch } from 'react-redux'
+import {
+    tSetUiProgram,
+    acUpdateUiProgramStage,
+    acClearUiProgram,
+} from '../../../actions/ui.js'
+import { useDebounce } from '../../../modules/utils.js'
+import {
+    sGetUiInputType,
+    sGetUiProgramId,
+    sGetUiProgramStage,
+} from '../../../reducers/ui.js'
+import { INPUT_TYPES } from '../InputPanel/index.js'
+import {
+    ProgramDimensionsFilter,
+    DIMENSION_TYPES,
+} from './ProgramDimensionsFilter.js'
+import { ProgramDimensionsList } from './ProgramDimensionsList.js'
+import styles from './ProgramDimensionsPanel.module.css'
+import { ProgramSelect } from './ProgramSelect.js'
+
+const PROGRAM_TYPES = {
+    WITHOUT_REGISTRATION: 'WITHOUT_REGISTRATION',
+    WITH_REGISTRATION: 'WITH_REGISTRATION',
+}
+const query = {
+    programs: {
+        resource: 'programs',
+        params: {
+            fields: [
+                'id',
+                'displayName',
+                'enrollmentDateLabel',
+                'incidentDateLabel',
+                'programType',
+                'programStages[id,displayName]',
+            ],
+            paging: false,
+        },
+    },
+}
+
+const ProgramDimensionsPanel = ({ visible }) => {
+    const dispatch = useDispatch()
+    const inputType = useSelector(sGetUiInputType)
+    const selectedProgramId = useSelector(sGetUiProgramId)
+    const selectedStageId = useSelector(sGetUiProgramStage)
+    const setSelectedProgramId = (id) => dispatch(tSetUiProgram(id))
+    const { fetching, error, data, refetch, called } = useDataQuery(query, {
+        lazy: true,
+    })
+    const [searchTerm, setSearchTerm] = useState('')
+    const [dimensionType, setDimensionType] = useState(DIMENSION_TYPES.ALL)
+    const debouncedSearchTerm = useDebounce(searchTerm, 350)
+    const filteredPrograms = data?.programs.programs.filter(
+        ({ programType }) =>
+            inputType === INPUT_TYPES.EVENT ||
+            programType === PROGRAM_TYPES.WITHOUT_REGISTRATION
+    )
+    const selectedProgram =
+        filteredPrograms &&
+        selectedProgramId &&
+        filteredPrograms.find(({ id }) => id === selectedProgramId)
```

Comment:
and potentially skip `selectedProgramId &&` unless there's a risk for matching `id` with `undefined`?

Reply:
> could maybe replace `filteredPrograms &&` with `filteredPrograms?.find`?

I did this

> and potentially skip `selectedProgramId &&` unless there's a risk for matching `id` with `undefined`?

I decided not to do this, since it expresses some intent: this very clearly signals that a selected program is never going to be set when there is no `selectedProgramId`. Omitting it introduces some mental overhead for simple people like me.

Also, it will prevent iterating over the `filteredPrograms` array when this is not needed, but we obviously don't have to start optimising for performance right now, so this is a minor reason.

Reply:
See bdfe9de